File: domain.go
package main

import (
	"errors"
	"sort"
	"time"

	"github.com/shopspring/decimal"
)

type LedgerEntry struct {
	ID       int64
	Kind     string
	Quantity decimal.Decimal
	Price    decimal.Decimal
	FXToIDR  decimal.Decimal
	At       time.Time
}

type Position struct {
	Quantity, AverageCost, CostBasis, Realized, NetInvested decimal.Decimal
}

func CalculatePosition(entries []LedgerEntry) (Position, error) {
	sort.SliceStable(entries, func(i, j int) bool {
		if entries[i].At.Equal(entries[j].At) {
			return entries[i].ID < entries[j].ID
		}
		return entries[i].At.Before(entries[j].At)
	})
	p := Position{}
	for _, e := range entries {
		if !e.Quantity.IsPositive() || !e.Price.IsPositive() || !e.FXToIDR.IsPositive() {
			return p, errors.New("quantity, price, and exchange rate must be positive")
		}
		value := e.Quantity.Mul(e.Price).Mul(e.FXToIDR)
		switch e.Kind {
		case "buy", "deposit":
			p.CostBasis = p.CostBasis.Add(value)
			p.Quantity = p.Quantity.Add(e.Quantity)
			p.AverageCost = p.CostBasis.Div(p.Quantity)
			p.NetInvested = p.NetInvested.Add(value)
		case "sell", "withdrawal":
			if e.Quantity.GreaterThan(p.Quantity) {
				return p, errors.New("quantity exceeds available holding")
			}
			cost := p.AverageCost.Mul(e.Quantity)
			p.Realized = p.Realized.Add(value.Sub(cost))
			p.CostBasis = p.CostBasis.Sub(cost)
			p.Quantity = p.Quantity.Sub(e.Quantity)
			p.NetInvested = p.NetInvested.Sub(value)
			if p.Quantity.IsZero() {
				p.AverageCost, p.CostBasis = decimal.Zero, decimal.Zero
			}
		default:
			return p, errors.New("invalid transaction type")
		}
	}
	return p, nil
}

func Convert(amount, usdToIDR decimal.Decimal, from, to string) (decimal.Decimal, error) {
	if from == to {
		return amount, nil
	}
	if !usdToIDR.IsPositive() {
		return decimal.Zero, errors.New("USD/IDR rate unavailable")
	}
	if from == "USD" && to == "IDR" {
		return amount.Mul(usdToIDR), nil
	}
	if from == "IDR" && to == "USD" {
		return amount.Div(usdToIDR), nil
	}
	return decimal.Zero, errors.New("unsupported currency")
}

File: sipd/domain.py
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal


@dataclass(frozen=True)
class LedgerEntry:
    id: int
    kind: str
    quantity: Decimal
    price: Decimal
    fx_to_idr: Decimal
    at: datetime


@dataclass
class Position:
    quantity: Decimal = Decimal()
    average_cost: Decimal = Decimal()
    cost_basis: Decimal = Decimal()
    realized: Decimal = Decimal()
    net_invested: Decimal = Decimal()


def calculate_position(entries: list[LedgerEntry]) -> Position:
    position = Position()
    for entry in sorted(entries, key=lambda value: (value.at, value.id)):
        if not all(value > 0 for value in (entry.quantity, entry.price, entry.fx_to_idr)):
            raise ValueError("quantity, price, and exchange rate must be positive")
        value = entry.quantity * entry.price * entry.fx_to_idr
        if entry.kind in {"buy", "deposit"}:
            position.cost_basis += value
            position.quantity += entry.quantity
            position.average_cost = position.cost_basis / position.quantity
            position.net_invested += value
        elif entry.kind in {"sell", "withdrawal"}:
            if entry.quantity > position.quantity:
                raise ValueError("quantity exceeds available holding")
            cost = position.average_cost * entry.quantity
            position.realized += value - cost
            position.cost_basis -= cost
            position.quantity -= entry.quantity
            position.net_invested -= value
            if not position.quantity:
                position.average_cost = position.cost_basis = Decimal()
        else:
            raise ValueError("invalid transaction type")
    return position


def convert(amount: Decimal, usd_to_idr: Decimal, source: str, target: str) -> Decimal:
    if source == target:
        return amount
    if usd_to_idr <= 0:
        raise ValueError("USD/IDR rate unavailable")
    if source == "USD" and target == "IDR":
        return amount * usd_to_idr
    if source == "IDR" and target == "USD":
        return amount / usd_to_idr
    raise ValueError("unsupported currency")
